src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -77,48 +112,110 @@ private void MakeInterpolatedStringFormat(BoundInterpolatedString node, out Boun
-                var builder = PooledStringBuilder.GetInstance();
-                var formatText = format.ConstantValue.StringValue;
-                int formatLength = formatText.Length;
-                for (int i = 0; i < formatLength; i++)
+                // All fill-ins, if any, are strings, and none of them have alignment or format specifiers.
+                // We can lower to a more efficient string concatenation
+                // The normal pattern for lowering is to lower subtrees before the enclosing tree. However in this case
+                // we want to lower the entire concatenation so we get the optimizations done by that lowering (e.g. constant folding).
+
+                int length = node.Parts.Length;
+                if (length == 0)
                 {
-                    char c = formatText[i];
-                    builder.Builder.Append(c);
-                    if ((c == '{' || c == '}') && (i + 1) < formatLength && formatText[i + 1] == c)
+                    // $"" -> ""
+                    return _factory.StringLiteral("");
+                }
+
+                if (length == 1)
+                {
+                    var part = node.Parts[0];
+
+                    if (part is BoundStringInsert fillin)
+                    {
+                        // $"{part}"
+                        // this is the filled-in expression
+                        part = fillin.Value;
+
+                        if (part.ConstantValue != null)
+                        {
+                            if (part.ConstantValue.StringValue == null)
+                            {
+                                return _factory.StringLiteral("");
+                            }
+                            else
+                            {
+                                result = part;
+                            }
+                        }
+                        else if (part is BoundBinaryOperator bbo && bbo.OperatorKind == BinaryOperatorKind.StringConcatenation)
+                        {
+                            // $"{a + b}" -> a + b
+                            result = part;
+                        }
+                        else
+                        {
+                            result = _factory.Coalesce(part, _factory.StringLiteral(""));
+                        }
+                    }
+                    else
+                    {
+                        // $"abc" -> "abc"
+                        // this is the literal part
+                        return _factory.StringLiteral(Unescape(part.ConstantValue.StringValue));
+                    }
+                }
+                else
+                {
+                    result = null;
+                    for (int i = 0; i < length; i++)
                     {
-                        i++;
+                        var part = node.Parts[i];
+                        if (part is BoundStringInsert fillin)
```

Comment:
I didn't understand why the treatment is different in the case where `length != 1` than it was in the case where `length == 1`.

I'd expect we can extract the logic from the `length == 1` into a method, then we can call the method in both cases and just check what's returned (if it's just a StringLiteral, in the case where `length == 1` then we can return). 
If that works out, then I suspect the check on `length == 1` will become a very local choice, rather than a containing `if` statement.

Something like:
```C#
foreach (var part in node.Parts)
{
    var loweredPart = LowerPart(part);
    if (length == 1 && /* loweredPart is a string literal */) { return loweredPart; }
    result = ...
}
```

Reply:
The reason why the `length == 1` case is different, is that it requires additional processing (e.g. lowering to a null coalescing expression) to ensure the resulting string is not null. Such processing is not required when `length > 1`, because string concatenation allready ensures it. I will add a comment to clarify that.

I did consider merging the loops.

Because of the additonal requirements for the `length == 1` case, such a `LowerPart` method would need two parameters: `LowerPart(BoundExpression part, bool isOnlyPart)`. While the loop duplication is not ideal, such a flag parameter isn't either. The parameter `isOnlyPart` is indeed tested locally in `LowerPart`, but has no local meaning. It becomes magic. And if you create two `LowerPart` variants, you might as well have the two loops.

As for the loop itself:

    bool isOnlyPart = length == 1;
    foreach (var part in node.Parts)
    {
        var loweredPart = LowerPart(part, isOnlyPart);
        if (isOnlyPart && loweredPart is BoundLiteral)
        {
            Debug.Assert(/* loweredPart is a non-null string*/);
            return loweredPart;
        }

        result = ...
    }

We can of course skip the if statement, and have the string literal go through another lowering pass like any other `result`. From a functional point of view, that doesn't hurt.

With or without the if statement, having a specialized loop for `length == 1` is more efficient.

Reply:
To clarify: the additional processing to ensure that the result is not null is _required_ when `length == 1`, but _forbidden_ when `length > 1`. The reason for that is that string concatenation does not flatten `((a + b) ?? "") + c` to `a + b + c`, and the same is true for `a + ((b + c) ?? "")`.

However, it just occurred to me that this problem can be fixed quite easily :-). I'll see what I can do.

Reply:
Done.